modules/dhcp_native_ms/dhcp_native_ms_main.rb

```diff
@@ -77,8 +78,27 @@ def unused_ip(subnet_address, mac_address, from_address, to_address)
       if dhcpsapi.api_level == DhcpsApi::Server::DHCPS_WIN2008_API || dhcpsapi.api_level == DhcpsApi::Server::DHCPS_NONE
         raise Proxy::DHCP::NotImplemented.new("DhcpsApi::Server#get_free_ip_address is not available on Windows Server 2008 and earlier versions.")
       end
-
-      return dhcpsapi.get_free_ip_address(subnet_address, from_address, to_address).first
+      
+      logger.debug "Searching for free IP in subnet #{subnet_address}"
+      while true
+        ip = dhcpsapi.get_free_ip_address(subnet_address, from_address, to_address).first
+        if ip.nil? || ip.empty?
+          logger.warn "No free IP returned by DHCP for subnet #{subnet_address}"
+          return nil
+        end
+        if system("ping -n 1 -w 1000 #{ip} > NUL")
```

Comment:
Please use tcp_pingable? and icmp_pingable? methods defined here: https://github.com/theforeman/smart-proxy/blob/develop/modules/dhcp_common/subnet.rb#L147. Probably should be extracted into a dedicated module or class.

Reply:
subnet.rb is not loaded anymore, do you want me to create a new module with only these 2 methods inside ?

Reply:
I think a module would be best.

Reply:
Please also check if tests need updating (most likely).

Reply:
@witlessbird I'm having troubles with the tcp_pingable method on windows (2012R2, ruby 2.4).
The timeout module does not work with TCPSocket.new:
```
puts Benchmark.measure {
  @timeout       = 3
  Timeout.timeout(@timeout) {
    sleep 5
  } rescue nil
}
  0.000000   0.000000   0.000000 (  3.001391)

puts Benchmark.measure {
  @timeout       = 3
  Timeout.timeout(@timeout) {
    TCPSocket.new("6.1.0.2", 7)
  } rescue nil
}
  0.000000   0.000000   0.000000 ( 21.003489)
```

Should I call only icmp_pingable for now?

Reply:
As we are no longer supporting rubies below 2.0, you can replace the above with something like:

```
Socket.tcp(ip_address, destination_port, timeout: timeout_in_seconds) 
```

Reply:
Ok thank you.

Reply:
The above you suggested works fine from Linux, but I have strange behaviour from Windows. If connect_timeout is specified, it throws ETIMEDOUT (after the specified timeout) instead of ECONNREFUSED:
```
irb(main):001:0> RUBY_PLATFORM
=> "x64-mingw32"
irb(main):002:0> RUBY_VERSION
=> "2.4.1"
irb(main):003:0> system("ping -n 1 -w 1000 172.16.0.1 > NUL")
=> true
irb(main):004:0> require 'socket'
=> true
irb(main):006:0> Socket.tcp('172.16.0.1', 7)
Errno::ECONNREFUSED: No connection could be made because the target machine actively refused it. - connect(2) for 172.16.0.1:7
...
irb(main):006:0> Socket.tcp('172.16.0.1', 7, connect_timeout: 5)
Errno::ETIMEDOUT: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond. - user specified timeout
...
```

Reply:
If you are up to it, you could rewrite this using select on tcp socket (which would allow to specify a timeout, and very likely will work ok).

Otherwise icmp check would probably suffice. Could you still move the tcp ping code into the same module/class that you moved icmp ping code into?

Reply:
Sure. My ruby skills are limited and I don't have much time to learn so I will go for option 2 :)